src/pldm_fw.cpp

```diff
@@ -7,33 +7,135 @@
 #include <libpldm/platform.h>
 #include <libpldm/pldm.h>
 #include <libpldm/state_set.h>
+#include <poll.h>
 
 #include <iostream>
 #include <sdbusplus/asio/connection.hpp>
 #include <sdbusplus/asio/object_server.hpp>
 
 namespace panel
 {
-types::Byte PldmFramework::getInstanceID()
+pldm_instance_id_t PldmFramework::getInstanceID()
 {
-    types::Byte instanceId = 0;
-    auto bus = sdbusplus::bus::new_default();
+    // Fetch Instance ID DB
+    int rc = pldm_instance_db_init_default(&pldmInstanceIdDb);
 
-    try
+    if (rc < 0)
+    {
+        throw std::runtime_error(
+            "Call to pldm_instance_db_init_default failed with return code " +
+            std::to_string(rc));
+    }
+
+    // Get Instance ID
+    pldm_instance_id_t instanceID = 0;
+
+    rc = pldm_instance_id_alloc(pldmInstanceIdDb, tid, &instanceID);
+    if (rc == -EAGAIN)
+    {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        rc = pldm_instance_id_alloc(pldmInstanceIdDb, tid, &instanceID);
+    }
+
+    if (rc < 0)
+    {
+        throw FunctionFailure(
+            "Call to pldm_instance_id_alloc failed with return code " +
+            std::to_string(rc));
+    }
+
+    std::cout << "Got instanceId: " << static_cast<int>(instanceID)
+              << " from PLDM eid: " << static_cast<int>(tid) << std::endl;
+    return instanceID;
+}
+
+void PldmFramework::freeInstanceID(pldm_instance_id_t instanceID)
+{
+    int rc = pldm_instance_id_free(pldmInstanceIdDb, tid, instanceID);
+    if (rc < 0)
+    {
+        std::cerr << "pldm_instance_id_free failed to free id = "
+                  << static_cast<int>(instanceID)
+                  << " of tid = " << static_cast<int>(tid) << " rc = " << rc
+                  << std::endl;
+    }
+
+    // Destroy the instance ID DB
+    rc = pldm_instance_db_destroy(pldmInstanceIdDb);
+    if (rc < 0)
+    {
+        std::cout << "pldm_instance_db_destroy failed, rc =" << rc << "\n";
+    }
+
+    if (pldmInstanceIdDb)
     {
-        auto method = bus.new_method_call(
-            "xyz.openbmc_project.PLDM", "/xyz/openbmc_project/pldm",
-            "xyz.openbmc_project.PLDM.Requester", "GetInstanceId");
-        method.append(mctpEid);
-        auto reply = bus.call(method);
-        reply.read(instanceId);
+        pldmInstanceIdDb = nullptr;
+        std::cout << "\n Destroyed pldm instance ID DB object manually"
+                  << std::endl;
     }
-    catch (const sdbusplus::exception::SdBusError& e)
+}
+
+int PldmFramework::openPldm()
+{
+    auto fd = -1;
+    if (pldmTransport)
     {
-        std::cerr << e.what() << std::endl;
-        throw FunctionFailure("pldm: call to GetInstanceId failed.");
+        std::cerr << "open: pldmTransport already setup!" << std::endl;
+        throw FunctionFailure("Required host dump action via PLDM is not "
+                              "allowed due to openPldm failed");
     }
-    return instanceId;
+
+    fd = openMctpDemuxTransport();
```

Comment:
Could have returned the function call itself. Don't see a need of another local variable.

Reply:
modified